lib/scholar/metrics/distance.ex

```diff
@@ -407,4 +426,62 @@ defmodule Scholar.Metrics.Distance do
     res = Nx.select(one_zero?, 0.0, res)
     1.0 - Nx.select(both_zero?, 1.0, res)
   end
+
+  @doc """
+  Hamming distance.
+
+  $hamming(x ,y) = \frac{\#\{x_{i, j...} \neq y_{i, j, ...}\}}{\#x_{i, j, ...}}$, where $i, j, ...$ are the aggregation axes
+
+  ## Options
+
+  #{NimbleOptions.docs(@hamming_schema)}
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 0, 0])
+      iex> y = Nx.tensor([0, 1, 0])
+      iex> Scholar.Metrics.Distance.hamming(x, y)
+      #Nx.Tensor<
+        f32
+        0.6666666865348816
+      >
+
+      iex> Scholar.Metrics.Distance.hamming(x, y, weights: [1,0.5,0.5])
+      #Nx.Tensor<
+        f32
+        0.75
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2, 3])
+      iex> Scholar.Metrics.Distance.hamming(x, y)
+      ** (ArgumentError) expected tensor to have shape {2}, got tensor with shape {3}
+
+      iex> x = Nx.tensor([[1, 2, 3], [0, 0, 0], [5, 2, 4]])
+      iex> y = Nx.tensor([[1, 5, 2], [2, 4, 1], [0, 0, 0]])
+      iex> Scholar.Metrics.Distance.hamming(x, y, axis: 1)
+      #Nx.Tensor<
+        f32[3]
+        [0.6666666865348816, 1.0, 1.0]
+      >
+  """
+  deftransform hamming(x, y, opts \\ []) do
+    options = NimbleOptions.validate!(opts, @hamming_schema)
+
+    if nil == options[:weights] do
+      hamming_unweighted(x, y, options)
+    else
+      hamming_weighted(x, y, options)
+    end
+  end
+
+  defnp hamming_unweighted(x, y, opts \\ []) do
+    assert_same_shape!(x, y)
+    (x != y) |> Nx.mean(axes: [opts[:axis]])
+  end
+
+  defnp hamming_weighted(x, y, opts \\ []) do
+    assert_same_shape!(x, y)
+    (x != y) |> Nx.weighted_mean(opts[:weights], axis: opts[:axis])
```

Comment:
I just realized we can't have `weights` as an option here.
This is because options are enclosed in most of our compilation infrastructure, so tensors must be passed as proper arguments to be kept as tensor expressions.

The code to deal with this is unfortunately a bit unwieldly, but it's basically like this:

```elixir
defmodule T do
import Nx.Defn

deftransform hamming(x, y), do: non_weighted(x, y)
deftransform hamming(x, y, opts) when is_list(opts) do
  # call nimble opts here :)
  non_weighted(x, y, opts)
end

deftransform hamming(x, y, w), do: weighted(x, y, w)
deftransform hamming(x, y, w, opts) when is_list(opts) do
  # call nimble opts here :)
  weighted(x, y, w, opts)
end

defnp weighted(x, y, w, opts \\ []) do
   ...
end

defnp non_weighted(x, y, opts \\ []) do
...
end
end
```

the guard on the second clause is what enables us to have both hamming/2 (non-weighted) with optional opts and hamming/3 (weighted) with optional opts

Reply:
I think i see what youre saying. I think weights make more sense as a tensor so that's probably the way to go rather than having them be a list and stay in options.

Reply:
Made the changes at 7f4c606fdd861c60237af92d416d48e34de21016

Using the new `weighted_mean` with `axes` parameter some other Scholar tests failed. Seems to be only the PCA tests

Reply:
The failed tests are actually due to SVD. I'll update Scholar's main branch with Nx main

Reply:
Done :)
You can update your fork with Scholar main and it'll pull latest Nx as a dependency

Reply:
Looks like I had to run mix format. I pushed again after pulling from Scholar/main